vnext/JSI.Desktop.UnitTests/JsiRuntimeUnitTests.cpp

```diff
@@ -0,0 +1,1056 @@
+//  EXPECT_TRUE(
+//    hasOwnPropertyName.call(rt, howpn, String::createFromAscii(rt, "false"))
+//    .getBool());
+//  EXPECT_TRUE(
+//    hasOwnPropertyName.call(rt, howpn, String::createFromAscii(rt, "3"))
+//    .getBool());
+//  EXPECT_TRUE(
+//    hasOwnPropertyName.call(rt, howpn, String::createFromAscii(rt, "c_prop"))
+//    .getBool());
+//  EXPECT_FALSE(hasOwnPropertyName
+//    .call(rt, howpn, String::createFromAscii(rt, "not_existing"))
+//    .getBool());
+//}
+
+TEST_P(JsiRuntimeUnitTests, ArrayTest) {
+  eval("x = {1:2, '3':4, 5:'six', 'seven':['eight', 'nine']}");
+
+  Object x = rt.global().getPropertyAsObject(rt, "x");
+  Array names = x.getPropertyNames(rt);
+  EXPECT_EQ(names.size(rt), 4);
+  std::unordered_set<std::string> strNames;
+  for (size_t i = 0; i < names.size(rt); ++i) {
+    Value n = names.getValueAtIndex(rt, i);
+    EXPECT_TRUE(n.isString());
+    strNames.insert(n.getString(rt).utf8(rt));
+  }
+
+  EXPECT_EQ(strNames.size(), 4);
+  EXPECT_EQ(strNames.count("1"), 1);
+  EXPECT_EQ(strNames.count("3"), 1);
+  EXPECT_EQ(strNames.count("5"), 1);
+  EXPECT_EQ(strNames.count("seven"), 1);
+
+  Object seven = x.getPropertyAsObject(rt, "seven");
+  Array arr = seven.getArray(rt);
+
+  EXPECT_EQ(arr.size(rt), 2);
+  EXPECT_EQ(arr.getValueAtIndex(rt, 0).getString(rt).utf8(rt), "eight");
+  EXPECT_EQ(arr.getValueAtIndex(rt, 1).getString(rt).utf8(rt), "nine");
+  // TODO: test out of range
+
+  EXPECT_EQ(x.getPropertyAsObject(rt, "seven").getArray(rt).size(rt), 2);
+
+  // Check that property access with both symbols and strings can access
+  // array values.
+  EXPECT_EQ(seven.getProperty(rt, "0").getString(rt).utf8(rt), "eight");
+  EXPECT_EQ(seven.getProperty(rt, "1").getString(rt).utf8(rt), "nine");
+  seven.setProperty(rt, "1", "modified");
+  EXPECT_EQ(seven.getProperty(rt, "1").getString(rt).utf8(rt), "modified");
+  EXPECT_EQ(arr.getValueAtIndex(rt, 1).getString(rt).utf8(rt), "modified");
+  EXPECT_EQ(
+    seven.getProperty(rt, PropNameID::forAscii(rt, "0"))
+    .getString(rt)
+    .utf8(rt),
+    "eight");
+  seven.setProperty(rt, PropNameID::forAscii(rt, "0"), "modified2");
+  EXPECT_EQ(arr.getValueAtIndex(rt, 0).getString(rt).utf8(rt), "modified2");
+
+  Array alpha = Array(rt, 4);
+  EXPECT_TRUE(alpha.getValueAtIndex(rt, 0).isUndefined());
+  EXPECT_TRUE(alpha.getValueAtIndex(rt, 3).isUndefined());
+  EXPECT_EQ(alpha.size(rt), 4);
+  alpha.setValueAtIndex(rt, 0, "a");
+  alpha.setValueAtIndex(rt, 1, "b");
+  EXPECT_EQ(alpha.length(rt), 4);
+  alpha.setValueAtIndex(rt, 2, "c");
+  alpha.setValueAtIndex(rt, 3, "d");
+  EXPECT_EQ(alpha.size(rt), 4);
+
+  EXPECT_TRUE(
+    function(
+      "function (arr) { return "
+      "arr.length == 4 && "
+      "['a','b','c','d'].every(function(v,i) { return v === arr[i]}); }")
+    .call(rt, alpha)
+    .getBool());
+
+  Array alpha2 = Array(rt, 1);
+  alpha2 = std::move(alpha);
+  EXPECT_EQ(alpha2.size(rt), 4);
+}
+
+TEST_P(JsiRuntimeUnitTests, FunctionTest) {
```

Comment:
This test and a few others are nearly verbatim from upstream, modulo new formatting changes. Can we use the upstream source files (e.g. via git submodule) and contribute any meaningful changes upstream? For tests that don't work on RN Windows/Chakra yet, we could redefine them via pre-processor (-DHostObjectTest=DISABLED_HostObjectTest). Compiling but not running is still valuable for knowing the ABI isn't diverging.

I'm concerned that diverging the test coverage efforts this way weakens RN as a whole.

Reply:
Could you please point me to the upstream source files you are referring to?

Reply:
I agree that consuming these tests directly from upstream is better, but I don't think that is possible right now due to upstream being private. Once upstream becomes public, I will make sure to revisit this area.

---
In reply to: [297374913](https://github.com/microsoft/react-native-windows/pull/2677#discussion_r297374913) [](ancestors = 297374913)